src/main/java/net/minecraftforge/resource/ResourceCacheManager.java

```diff
@@ -0,0 +1,406 @@
+         * @param entry      The entry to inject into the cache, must not be null.
+         */
+        private void injectIntoCache(@Nullable final Path parentPath, final ResourceCacheEntry entry)
+        {
+            String pathEntry;
+            if (parentPath == null)
+            {
+                // We have null, generally this only happens on the root directory.
+                pathEntry = "";
+            } else
+            {
+                // Use the parent paths, we can not depend on a toString call here, since we need the / as separator.
+                pathEntry = Joiner.on("/").join(parentPath);
+            }
+
+            // Inject into the cache.
+            this.entriesByPathPrefix.computeIfAbsent(pathEntry, e -> new CopyOnWriteArrayList<>()).add(entry);
+
+            // Recursively walk to the top, while preventing duplicate entries.
+            if (parentPath != null && !pathEntry.isEmpty())
+            {
+                this.injectIntoCache(parentPath.getParent(), entry);
+            }
+        }
+
+        /**
+         * Looks up resources in the given input path prefix as well as those that match the filter.
+         *
+         * @param inputPath The input path prefix to look in.
+         * @param filter    The filter which the resources must match.
+         * @return A collection of resource location which match the given path prefix and who match the filter.
+         */
+        public Collection<ResourceLocation> getResources(final Path inputPath, final Predicate<ResourceLocation> filter)
+        {
+            // Use the input path prefix, we can not depend on a toString call here, since we need the / as separator.
+            final String pathEntry = Joiner.on('/').join(inputPath);
+
+            // Since we inject into the cache recursively we can now just grab the map entry that is the prefix and loop over its values.
+            // We use getOrDefault with a stream combination since the returned list needs to be mutable.
+            return entriesByPathPrefix.getOrDefault(pathEntry, Collections.emptyList()).stream()
+                    .map(ResourceCacheEntry::resourceLocation)
+                    .filter(filter)
+                    .collect(Collectors.toList());
+        }
+
+        /**
+         * Indicates if the cache of this manager has been loaded.
+         *
+         * @return True if the cache has been loaded, false otherwise.
+         */
+        public boolean cacheLoaded()
+        {
+            return cacheLoaded.get();
+        }
+    }
+
+    /**
+     * Functional callback interface to get a walkable stream of paths.
+     * Supports throwing {@link IOException} if the stream can not be created.
+     */
+    @FunctionalInterface
+    private interface PathWalkerFactory
+    {
+        Stream<Path> createWalkingStream(Path path) throws IOException;
+    }
```

Comment:
It might be needed in the future, right now we handle Path and Vanilla only as packs we adapt, but folder and file might also gain benefits from this treatment, although I was not able to completely verify this, which is why this abstraction layer exists for now, and when I or somebody else have more time this system can be extended.